primitives/primproc/bppsendthread.cpp

```diff
@@ -53,29 +51,38 @@ BPPSendThread::~BPPSendThread()
 {
     boost::mutex::scoped_lock sl(msgQueueLock);
     boost::mutex::scoped_lock sl2(ackLock);
+    boost::mutex::scoped_lock sl3(respondLock);
     die = true;
     queueNotEmpty.notify_one();
     okToSend.notify_one();
+    okToRespond.notify_one();
     sl.unlock();
     sl2.unlock();
+    sl3.unlock();
     runner.join();
 }
 
-bool BPPSendThread::okToProceed()
-{
-    // keep the queue size below the 100 msg threshold & below the 25MB mark,
-    // but at least 2 msgs so there is always 1 ready to be sent.
-    return ((msgQueue.size() < sizeThreshold && currentByteSize < maxByteSize)
-            || msgQueue.size() < 3) && !die;
-}
-
 void BPPSendThread::sendResult(const Msg_t& msg, bool newConnection)
 {
+    // Wait for the queue to empty out a bit if it's stuffed full
+    if (sizeTooBig())
+    {
+        boost::mutex::scoped_lock sl1(respondLock);
+        while (currentByteSize >= maxByteSize && msgQueue.size() > 3 && !die)
```

Comment:
We have this condition in multiple places. Changing one of them it is easy to forget the others.

Reply:
Moreover we have the same scope in the parenthesis in two BPPSendThread methods so it is better to wrap it up in a function.

Reply:
It could be even harder to read if all this is in a seperate function, though I could do that. However, after a code search. the function sendResults() appears to no longer be used anywhere. I don't know if we would ever want to resurrect it. Perhaps if we just delete the unreferenced function?

Reply:
As for the two remaining places where sizeTooBig is called, in sendResult() this is blocking the thread since it's in the middle of a block. In BatchPrimitiveProcessor::operator(), a positive too big will result in the BPP being returned to the pool and the job rescheduled.

Reply:
I think it is a good idea to comment out/remove unreferenced function.

Reply:
I created MCOL-4969 to remove SendResults.